echo-scheduler/src/main/groovy/com/netflix/spinnaker/echo/scheduler/actions/pipeline/PipelineConfigsPollingJob.groovy

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2015 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.echo.scheduler.actions.pipeline
+
+import com.netflix.spectator.api.Registry
+import com.netflix.spinnaker.echo.model.Trigger
+import com.netflix.spinnaker.echo.pipelinetriggers.PipelineCache
+import groovy.util.logging.Slf4j
+import org.quartz.CronTrigger
+import org.quartz.DisallowConcurrentExecution
+import org.quartz.Job
+import org.quartz.JobExecutionContext
+import org.quartz.Scheduler
+import org.quartz.TriggerKey
+import org.quartz.impl.matchers.GroupMatcher
+
+import java.util.concurrent.TimeUnit
+
+/**
+ * Syncs triggers from pipelines with the triggers in the scheduler
+ */
+@Slf4j
+@DisallowConcurrentExecution
+class PipelineConfigsPollingJob implements Job {
+  private TimeZone timeZoneId
+  private Scheduler scheduler
+  private Registry registry
+  private PipelineCache pipelineCache
+
+  @SuppressWarnings("unused")
+  PipelineConfigsPollingJob() {
+    pipelineCache = (PipelineCache) SchedulerBeanDependencies.getBean(PipelineCache)
+    registry = (Registry) SchedulerBeanDependencies.getBean(Registry)
+  }
+
+  PipelineConfigsPollingJob(Registry registry, PipelineCache pipelineCache) {
+    this.pipelineCache = pipelineCache
+    this.registry = registry
+  }
+
+  void execute(JobExecutionContext context) {
+    long start = System.currentTimeMillis()
+
+    try {
+      log.info("Running the pipeline configs polling agent...")
+
+      timeZoneId = TimeZone.getTimeZone(context.mergedJobDataMap.getString("timeZoneId"))
+      scheduler = context.scheduler
+
+      // Only interested in pipelines that have enabled CRON triggers
+      def pipelinesWithCronTriggers = pipelineCache.getPipelinesSync().findAll { pipeline ->
+        !pipeline.disabled && pipeline.triggers && pipeline.triggers.enabled &&
+          pipeline.triggers.any { Trigger.Type.CRON.toString().equalsIgnoreCase(it.type) }
+      }
+
+      log.debug("Found ${pipelinesWithCronTriggers.size()} pipeline CRON triggers that are active")
+
+      def pipelineTriggers = new TriggerRepository(pipelinesWithCronTriggers)
+
+      removeStaleTriggers(pipelineTriggers)
+      updateChangedTriggers(pipelineTriggers)
+
+      registry.gauge("trigger.count").set(pipelinesWithCronTriggers.size())
+    } catch (Exception e) {
+      log.error("Exception occurred in the execute() method of PipelineConfigsPollingAgent", e)
```

Comment:
I'd change this to read more like "Error occurred while syncing cron pipelines", since the stacktrace will give the location of the error.

Reply:
👍